core/trino-main/src/test/java/io/trino/type/TestDecimalOperators.java

```diff
@@ -280,43 +318,75 @@ public void testModulus()
         assertDecimalFunction("DECIMAL '7.000000000000000000000000000000000000' % DECIMAL '3'", decimal("1.000000000000000000000000000000000000"));
         assertDecimalFunction("DECIMAL '-7.000000000000000000000000000000000000' % DECIMAL '3'", decimal("-1.000000000000000000000000000000000000"));
         assertDecimalFunction("DECIMAL '7.000000000000000000000000000000000000' % DECIMAL '-3'", decimal("1.000000000000000000000000000000000000"));
         assertDecimalFunction("DECIMAL '-7.000000000000000000000000000000000000' % DECIMAL '-3'", decimal("-1.000000000000000000000000000000000000"));
+        assertDecimalFunction("DECIMAL '9.000000000000000000000000000000000000' % DECIMAL '3'", decimal("0.000000000000000000000000000000000000"));
+        assertDecimalFunction("DECIMAL '-9.000000000000000000000000000000000000' % DECIMAL '3'", decimal("0.000000000000000000000000000000000000"));
+        assertDecimalFunction("DECIMAL '9.000000000000000000000000000000000000' % DECIMAL '-3'", decimal("0.000000000000000000000000000000000000"));
+        assertDecimalFunction("DECIMAL '-9.000000000000000000000000000000000000' % DECIMAL '-3'", decimal("0.000000000000000000000000000000000000"));
 
         // long long -> long
+        assertDecimalFunction("CAST(0 AS DECIMAL(38,0)) % CAST(3 AS DECIMAL(38,0))", decimal("00000000000000000000000000000000000000"));
+        assertDecimalFunction("CAST(0 AS DECIMAL(38,0)) % CAST(-3 AS DECIMAL(38,0))", decimal("00000000000000000000000000000000000000"));
         assertDecimalFunction("CAST(7 AS DECIMAL(38,0)) % CAST(3 AS DECIMAL(38,0))", decimal("00000000000000000000000000000000000001"));
         assertDecimalFunction("CAST(7 AS DECIMAL(34,0)) % CAST(3 AS DECIMAL(38,0))", decimal("0000000000000000000000000000000001"));
         assertDecimalFunction("CAST(7 AS DECIMAL(38,0)) % CAST(3 AS DECIMAL(34,0))", decimal("0000000000000000000000000000000001"));
         assertDecimalFunction("CAST(-7 AS DECIMAL(38,0)) % CAST(3 AS DECIMAL(38,0))", decimal("-00000000000000000000000000000000000001"));
         assertDecimalFunction("CAST(7 AS DECIMAL(38,0)) % CAST(-3 AS DECIMAL(38,0))", decimal("00000000000000000000000000000000000001"));
         assertDecimalFunction("CAST(-7 AS DECIMAL(38,0)) % CAST(-3 AS DECIMAL(38,0))", decimal("-00000000000000000000000000000000000001"));
+        assertDecimalFunction("CAST(9 AS DECIMAL(38,0)) % CAST(3 AS DECIMAL(38,0))", decimal("00000000000000000000000000000000000000"));
+        assertDecimalFunction("CAST(-9 AS DECIMAL(38,0)) % CAST(3 AS DECIMAL(38,0))", decimal("00000000000000000000000000000000000000"));
+        assertDecimalFunction("CAST(9 AS DECIMAL(38,0)) % CAST(-3 AS DECIMAL(38,0))", decimal("00000000000000000000000000000000000000"));
+        assertDecimalFunction("CAST(-9 AS DECIMAL(38,0)) % CAST(-3 AS DECIMAL(38,0))", decimal("00000000000000000000000000000000000000"));
```

Comment:
should you add `-0 % xxx` case?

Reply:
I don't think tests that take negative zero as input have sense. We're only concerned with not getting it as output.